### PR TITLE
Feature/pm 509 interpolate subgaits with parameter

### PR DESCRIPTION
**Pull request by RutgerVanBeek**
Closes PM-509

## Description

We want to able to interpolate two subgaits with a parameter to obtain a new subgait. This PR implements methods in the Subgait, JointTrajectory and Setpoint classes to do so. I also added a function which can load two subgait files and interpolate them. The functions are not used currrently, but in a new issue the 'gait selection', 'default.yaml' and the 'rqt_gait_selection' should be updated to make use of these functions. The functions are tested though as I wrote tests for the Subgait, JointTrajectory, Setpoint and Limits classes.
## Changes

* Added interpolation methods for Subgait, JointTrajectory and Setpoint classes
* Added (un)equals operator for the limits
* Added function to load two subgaits and interpolate them
* Added 'SubgaitInterpolationError' for when subgaits differed to much to interpolate
* Added tests to test the new functionality.

### CODE REVIEW

File: march_shared_classes/src/march_shared_classes/exceptions/gait_exceptions.py [hunk 1]
```diff
@@ -61,3 +61,12 @@ def __init__(self, msg=None):
             msg = 'Subgaits can not transition'
 
         super(TransitionError, self).__init__(msg)
+
+
+class SubgaitInterpolationError(Exception):
+    def __init__(self, msg=None):
+        """Class to raise an error when it was not possible to interpolate between subgaits."""
+        if msg is None:
+            msg = 'An error occurred while trying to merge two subgaits.'
```
Comment (JorisWeeda) [inline march_shared_classes/src/march_shared_classes/exceptions/gait_exceptions.py:70]: maybe nice to add which subgaits for debugging?

File: march_shared_classes/src/march_shared_classes/exceptions/gait_exceptions.py [hunk 2]
```diff
@@ -61,3 +61,12 @@ def __init__(self, msg=None):
             msg = 'Subgaits can not transition'
 
         super(TransitionError, self).__init__(msg)
+
+
+class SubgaitInterpolationError(Exception):
```
Comment (JorisWeeda) [inline march_shared_classes/src/march_shared_classes/exceptions/gait_exceptions.py:66]:
```suggestion
class JointInterpolationError(Exception):
```

File: march_shared_classes/test/subgait_test.py [hunk 3]
```diff
@@ -140,3 +153,70 @@ def test_equalize_amount_of_setpoints_with_lower_duration_new_gait(self):
...
+        base_subgait, other_subgait = self.load_interpolatable_subgaits('right_close', 'MV_walk_rightclose_v2',
+                                                                        'MV_walk_rightclose_v2_seven_joints')
+        with self.assertRaises(SubgaitInterpolationError):
+            Subgait.interpolate_subgaits(base_subgait, other_subgait, 0.5)
+
+    def test_interpolate_subgaits_wrong_joint_names(self):
+        base_subgait, other_subgait = self.load_interpolatable_subgaits(
+            other_version='MV_walk_leftclose_v2_wrong_joint_name')
+        with self.assertRaises(SubgaitInterpolationError):
+            Subgait.interpolate_subgaits(base_subgait, other_subgait, 0.5)
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py:215]: I think this test is better moved to joint_trajectory_test

File: march_shared_classes/test/subgait_test.py [hunk 4]
```diff
@@ -140,3 +153,70 @@ def test_equalize_amount_of_setpoints_with_lower_duration_new_gait(self):
...
+
+    def test_interpolate_subgaits_parameter_one(self):
+        base_subgait, other_subgait = self.load_interpolatable_subgaits()
+        new_subgait = Subgait.interpolate_subgaits(base_subgait, other_subgait, 1)
+        self.assertEqual(other_subgait, new_subgait)
+
+    def test_interpolate_subgaits_interpolated(self):
+        # test whether each setpoint of the new subgait is between the setpoins
+        base_subgait, other_subgait = self.load_interpolatable_subgaits()
+        new_subgait = Subgait.interpolate_subgaits(base_subgait, other_subgait, 1)
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py:191]:
```suggestion
        new_subgait = Subgait.interpolate_subgaits(base_subgait, other_subgait, 0.4)
```
You could even go so far as to input a random value between 0 and 1 here

File: march_shared_classes/test/subgait_test.py [hunk 5]
```diff
@@ -35,6 +35,19 @@ def test_from_file_none_path(self):
     def test_from_file_no_robot(self):
         self.assertIsNone(Subgait.from_file(None, self.subgait_path))
 
+    # Subgait.from_files_interpolated tests
+    def test_from_files_interpolated_correct(self):
+        base_file_name = '{rsc}/{gait}/{subgait}/{version}.subgait'.format(rsc=self.resources_folder,
+                                                                           gait=self.gait_name,
+                                                                           subgait='left_close',
+                                                                           version='MV_walk_leftclose_v1')
+        other_file_name = '{rsc}/{gait}/{subgait}/{version}.subgait'.format(rsc=self.resources_folder,
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py:44]:
```suggestion
        base_subgait_path = '{rsc}/{gait}/{subgait}/{version}.subgait'.format(rsc=self.resources_folder,
                                                                           gait=self.gait_name,
                                                                           subgait='left_close',
                                                                           version='MV_walk_leftclose_v1')
        other_subgait_path = '{rsc}/{gait}/{subgait}/{version}.subgait'.format(rsc=self.resources_folder,
```
Your naming is not consistent with the method below ;)

File: march_shared_classes/test/subgait_test.py [hunk 6]
```diff
@@ -140,3 +153,70 @@ def test_equalize_amount_of_setpoints_with_lower_duration_new_gait(self):
...
+            other_version='MV_walk_leftclose_v2_wrong_joint_name')
+        with self.assertRaises(SubgaitInterpolationError):
+            Subgait.interpolate_subgaits(base_subgait, other_subgait, 0.5)
+
+    def test_interpolate_subgaits_duration(self):
+        base_subgait, other_subgait = self.load_interpolatable_subgaits()
+        parameter = 0.2
+        new_subgait = Subgait.interpolate_subgaits(base_subgait, other_subgait, parameter)
+        new_duration = parameter * base_subgait.duration + (1 - parameter) * other_subgait.duration
+        self.assertEqual(new_duration, new_subgait.duration)
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py:222]: Hmm maybe you should also write such a test for joint_trajectory

File: march_shared_classes/test/subgait_test.py [hunk 7]
```diff
@@ -140,3 +153,70 @@ def test_equalize_amount_of_setpoints_with_lower_duration_new_gait(self):
...
+                base_setpoint = base_subgait.joints[i].setpoints[j]
+                other_setpoint = other_subgait.joints[i].setpoints[j]
+                self.assertTrue(min(base_setpoint.time, other_setpoint.time) <= setpoint.time
+                                <= max(base_setpoint.time, other_setpoint.time))
+
+                self.assertTrue(min(base_setpoint.position, other_setpoint.position) <= setpoint.position
+                                <= max(base_setpoint.position, other_setpoint.position))
+
+                self.assertTrue(min(base_setpoint.velocity, other_setpoint.velocity) <= setpoint.velocity
+                                <= max(base_setpoint.velocity, other_setpoint.velocity))
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py:203]: You could even full calculate the supposed value here instead of just testing that it is in between these values.

File: march_shared_classes/test/limits_test.py [hunk 8]
```diff
@@ -24,3 +24,10 @@ def test_k_position(self):
 
     def test_k_velocity(self):
         self.assertEqual(self.limits.k_velocity, 5, 'k_velocity not initialised correctly.')
+
+    def test_equal_operator(self):
+        self.assertEqual(self.limits, Limits(0, 1, 2, 3, 4, 5))
+
+    def test_unequal_operator(self):
+        # first entry different
+        self.assertNotEqual(self.limits, Limits(1, 1, 2, 3, 4, 5))
```
[the remark this answers is not in the document]
Comment (RutgerVanBeek) [reply march_shared_classes/test/limits_test.py:33]: Doubted about this, that it was a bit trivial, but is indeed more neater.

File: march_shared_classes/test/setpoint_test.py [hunk 9]
```diff
@@ -35,6 +35,13 @@ def test_unequal_position(self):
...
-    def test_eunqual_velocity(self):
+    def test_unequal_velocity(self):
         other_setpoint = Setpoint(1.1234, 0.0343, 0)
         self.assertNotEqual(self.setpoint, other_setpoint)
+
+    def test_interpolation_correct(self):
+        parameter = 0.3
+        other_setpoint = Setpoint(1, 1, 1)
+        expected_result = Setpoint(parameter * 1.1234 + (1 - parameter) * 1, 0.03434 * parameter + (1 - parameter) * 1,
+                                   123.1620 * parameter + (1 - parameter) * 1)
```
[the remark this answers is not in the document]
Comment (RutgerVanBeek) [reply march_shared_classes/test/setpoint_test.py:46]: I think it is cleaner, it more easier to see the consistency this way.

File: march_shared_classes/test/subgait_test.py (hunk 4 above)
Comment (RutgerVanBeek) [reply march_shared_classes/test/subgait_test.py:191]: There is this packages hypothesis, which can do this very well and searches for edge cases.

File: march_shared_classes/test/subgait_test.py (hunk 3 above)
Comment (RutgerVanBeek) [reply march_shared_classes/test/subgait_test.py:215]: > One thought: Since these methods only get used in the gait selection, it thought about creating a seperate subclass for this (like InterpolatedSubgait or something). That class would be a bit empty. I think it would make these parent classes clearer at the cost of some vagueness due to the extra file. Alternatively, it could also be a part of that interface refactor we were talking about

The result is just a normal subgait, with which you can interact as with a normal subgait. I think adding an extra class does not make it extra clear. Why might split it to multiple files though.

**Comment by RutgerVanBeek** [comment]
> One thought: Since these methods only get used in the gait selection, it thought about creating a seperate subclass for this (like InterpolatedSubgait or something). That class would be a bit empty. I think it would make these parent classes clearer at the cost of some vagueness due to the extra file. Alternatively, it could also be a part of that interface refactor we were talking about

The result is just a normal subgait, with which you can interact as with a normal subgait. I think adding an extra class does not make it extra clear. Why might split it to multiple files though.

File: march_shared_classes/test/setpoint_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RutgerVanBeek) [reply march_shared_classes/test/setpoint_test.py:46]: I think it is more clear, when the *1 is included.

File: march_shared_classes/test/subgait_test.py (hunk 3 above)
Comment (RutgerVanBeek) [reply march_shared_classes/test/subgait_test.py:215]: I moved the name test to the method in the Subgait class, because of the usage of the get_joint method. Therefore the test now belongs here. (But you were right initially).

File: march_shared_classes/test/joint_trajectory_test.py [hunk 10]
```diff
@@ -86,3 +87,15 @@ def test_get_interpolated_setpoints_home_subgait(self):
...
+        different_limits = Limits(-2, 3, 2)
+        other_trajectory = JointTrajectory(self.joint_name, different_limits, self.setpoints, self.duration)
+        with self.assertRaises(SubgaitInterpolationError):
+            JointTrajectory.interpolate_joint_trajectories(self.joint_trajectory, other_trajectory, 0.5)
+
+    def test_interpolate_trajectories_unequal_amount_setpoints(self):
+        other_trajectory = JointTrajectory(self.joint_name, self.limits, [self.setpoints[0], self.setpoints[-1]],
+                                           self.duration)
+        with self.assertRaises(SubgaitInterpolationError):
+            JointTrajectory.interpolate_joint_trajectories(self.joint_trajectory, other_trajectory, 0.5)
```
[the remark this answers is not in the document]
Comment (RutgerVanBeek) [reply march_shared_classes/test/joint_trajectory_test.py:101]: I move the name check to the method in the subgait class and there it is already checked, so I will not do this.

File: march_shared_classes/test/subgait_test.py (hunk 4 above)
Comment (RutgerVanBeek) [reply march_shared_classes/test/subgait_test.py:191]: I agree that 1 is not the best value here, although I think random is not the best practice in testing. We might consider using the hypothesis package if we want to test this more generally.

File: march_shared_classes/test/subgait_test.py [hunk 11]
```diff
@@ -187,20 +187,18 @@ def test_interpolate_subgaits_parameter_one(self):
...
-                self.assertTrue(min(base_setpoint.position, other_setpoint.position) <= setpoint.position
-                                <= max(base_setpoint.position, other_setpoint.position))
-
-                self.assertTrue(min(base_setpoint.velocity, other_setpoint.velocity) <= setpoint.velocity
-                                <= max(base_setpoint.velocity, other_setpoint.velocity))
+                self.assertTrue(base_setpoint.time * parameter + (1 - parameter) * other_setpoint.time, setpoint.time)
+                self.assertTrue(base_setpoint.position * parameter + (1 - parameter) * other_setpoint.position,
+                                setpoint.position)
+                self.assertTrue(base_setpoint.velocity * parameter + (1 - parameter) * other_setpoint.velocity,
+                                setpoint.velocity)
```
Comment (Roelemans) [inline march_shared_classes/test/subgait_test.py]: :joy:

File: march_shared_classes/src/march_shared_classes/gait/joint_trajectory.py [hunk 12]
```diff
@@ -140,3 +142,17 @@ def __getitem__(self, index):
 
     def __len__(self):
         return len(self.setpoints)
+
+    @classmethod
+    def interpolate_joint_trajectories(cls, base_trajectory, other_trajectory, parameter):
```
Comment (JorisWeeda) [inline march_shared_classes/src/march_shared_classes/gait/joint_trajectory.py:147]: i would add a docstring to the function to explain parameter or rename the variable to _transition_parameter_ (or something like that)

File: march_shared_classes/src/march_shared_classes/gait/setpoint.py [hunk 13]
```diff
@@ -46,3 +46,10 @@ def __eq__(self, other):
 
     def __ne__(self, other):
         return not self.__eq__(other)
+
+    @staticmethod
+    def interpolate_setpoints(base_setpoint, other_setpoint, parameter):
```
Comment (JorisWeeda) [inline march_shared_classes/src/march_shared_classes/gait/setpoint.py:51]: same here for the parameter variable

File: march_shared_classes/test/setpoint_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Roelemans) [reply march_shared_classes/test/setpoint_test.py:46]: I disagree with the values: If you were to change them at the start of the file, this test would break.

File: march_shared_classes/src/march_shared_classes/gait/subgait.py [hunk 14]
```diff
@@ -229,6 +235,37 @@ def create_interpolated_setpoints(self, timestamps):
...
+    def interpolate_subgaits(cls, base_subgait, other_subgait, parameter):
+        if parameter == 1:
+            return other_subgait
+        if parameter == 0:
+            return base_subgait
+        if not (0 < parameter < 1):
+            raise ValueError('Parameter for interpolation should be in the interval [0, 1], but is {0}'
+                             .format(parameter))
+
+        if len(base_subgait.joints) != len(other_subgait.joints):
```
Comment (JorisWeeda) [inline march_shared_classes/src/march_shared_classes/gait/subgait.py:248]: wouldn't it be better to:
```suggestion
        if base_subgait.joints == other_subgait.joints:
```
or could a possible situation be that the joints are not in the same order, in that case:
```suggestion
        if sorted(base_subgait.joints) == sorted(other_subgait.joints):
```

File: march_shared_classes/test/setpoint_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (RutgerVanBeek) [reply march_shared_classes/test/setpoint_test.py:46]: I changed it, but have you seen the other tests in this file?

File: march_shared_classes/test/setpoint_test.py [hunk 15]
```diff
@@ -42,6 +42,7 @@ def test_unequal_velocity(self):
     def test_interpolation_correct(self):
         parameter = 0.3
         other_setpoint = Setpoint(1, 1, 1)
-        expected_result = Setpoint(parameter * 1.1234 + (1 - parameter) * 1, 0.03434 * parameter + (1 - parameter) * 1,
-                                   123.1620 * parameter + (1 - parameter) * 1)
+        expected_result = Setpoint(parameter * self.setpoint.time + (1 - parameter) * 1,
+                                   self.setpoint.position * parameter + (1 - parameter) * 1,
+                                   self.setpoint.velocity * parameter + (1 - parameter) * 1)
```
Comment (Roelemans) [inline march_shared_classes/test/setpoint_test.py:47]:
```suggestion
        expected_result = Setpoint(self.setpoint.time * parameter + (1 - parameter) * 1,
                                   self.setpoint.position * parameter + (1 - parameter) * 1,
                                   self.setpoint.velocity * parameter + (1 - parameter) * 1)
```
